### DISCUSSION TITLE
Reduce spacy package size for AWS Lambda limits

### DISCUSSION THREAD

**Discussion by MrRobotV8** [explosion/spaCy#11783] [Help: Best practices]
Hi!

I would like to perform NER with spacy en_core_web_sm into an AWS Lambda which has already other packages installed (like scrapy). 

Is there a way to reduce the installation size of the model? 

The usage is: 
```
nlp = spacy.load(EN_CORE_WEB_SM + '/en_core_web_sm-3.4.1/')
doc = nlp(text)
```
Logic on doc.ents. 

I'm creating a Layer from a Docker File where I installed scrapy and spaCy. 

requirements.txt
spacy
scrapy
fuzzywuzzy
python-Levenshtein
python-dateutil

the en_core_web_sm can be downloaded from S3 in order to create a smaller layer. But I woul like to reduce the size of the spacy installation if it's possible. 

```
FROM public.ecr.aws/sam/build-python3.9
ARG version=3.9
WORKDIR /build
COPY requirements.txt ./

RUN mkdir -p python/lib/python$version/site-packages/
RUN pip install -U pip setuptools wheel
RUN pip install -r ./requirements.txt -t python/lib/python$version/site-packages/



# # Setup spacy again in system path to execute spacy downloader
RUN pip install spacy
# Setup directory for spacy_DATA


# Download spacy_DATA to build directory
RUN python -m spacy download en_core_web_sm -t /build/

RUN mkdir -p ./share
RUN zip -r ./share/python-spacy-layer.zip python en_core_web_sm

```
Thank you for your help!

**Comment by adrianeboyd** [comment]
`en_core_web_sm` is already relatively small, only ~15MB. If you only need `ner`, you can reduce the size to ~8MB by removing all components except `"ner"` with `nlp.remove_pipe()` and then saving this with `nlp.to_disk` to a custom path. It's not really recommended but if you really need to save another 1M, you can replace `vocab/strings.json` with an empty JSON list `[]`.

There are other places to check, too, for instance it currently looks like using `--no-binary pydantic` saves nearly 50MB in linux in the pydantic install: https://pydantic-docs.helpmanual.io/install/#performance-vs-package-size-trade-off. (It also looks like they're considering improving the default builds so maybe pydantic v1.10.3 won't be so huge.)

**Reply by MrRobotV8** [reply]
Thank you! The en_core_web_sm is not a big deal. The issue is the **spacy** package.. I need to install the spacy package to use the model and when you run the command, the package installation results in several dependencies that I need to have on the Lambda Function. 
Just for Spice the site-packages of my VirtualEnv are about 180Mb Unzipped. 
Immagine to add 15 Mb for en_core_web_sm, and the other dependencies from other requirements, it result in something larger than 250Mb, which is the Lambda Limit.

**Reply by adrianeboyd** [reply]
So you should definitely try the python-only pydantic installation.

I don't know about your other dependencies and how much they depend on numpy, but the next not-exactly-recommend space-saving option would be to compile numpy without any additional BLAS libraries and optimized for size (instead of speed or debugging, similar to the pydantic recommendations if you use the compiled version):

```
pip uninstall -y numpy
CFLAGS="-Os -g0 -s" NPY_BLAS_ORDER= NPY_LAPACK_ORDER= pip install --no-binary numpy --no-cache-dir numpy
```

That saves ~35MB in linux for me but can be a lot slower depending on what you're doing. spacy mainly uses `blis` instead of `numpy` for matrix multiplication in the models so it doesn't look like it's that much slower for NER, but other spacy functionality (similarity is the main one that immediately comes to mind) or non-spacy packages could be affected.

So all three things would save almost 90MB, and beyond that you'll just need to look at what's installed and search for advice like for pydantic about how to compile a package with different options or install an alternate smaller version.

The numpy option has the potential to get iffy in terms of speed, so the pydantic install is the best place to start.
